sql-plugin/src/main/spark320/scala/com/nvidia/spark/rapids/shims/HybridFileSourceScanExecMeta.scala

```diff
@@ -0,0 +1,188 @@
+            case _ =>
+              dpe
+          }
+      }
+    }
+  }
+
+  // partition filters and data filters are not run on the GPU
+  override val childExprs: Seq[ExprMeta[_]] = Seq.empty
+
+  override def tagPlanForGpu(): Unit = {
+    val cls = wrapped.relation.fileFormat.getClass
+    if (cls != classOf[ParquetFileFormat]) {
+      willNotWorkOnGpu(s"unsupported file format: ${cls.getCanonicalName}")
+    }
+  }
+
+  override def convertToGpu(): GpuExec = {
+    // Modifies the original plan to support DPP
+    val fixedExec = wrapped.copy(partitionFilters = partitionFilters)
+    HybridFileSourceScanExec(fixedExec)(conf)
+  }
+}
+
+object HybridFileSourceScanExecMeta {
+  // Determines whether using HybridScan or GpuScan
+  def useHybridScan(conf: RapidsConf, fsse: FileSourceScanExec): Boolean = {
+    val isEnabled = if (conf.useHybridParquetReader) {
+      require(conf.loadHybridBackend,
+        "Hybrid backend was NOT loaded during the launch of spark-rapids plugin")
+      true
+    } else {
+      false
+    }
+    // Currently, only support reading Parquet
+    lazy val isParquet = fsse.relation.fileFormat.getClass == classOf[ParquetFileFormat]
+    // Check if data types of all fields are supported by HybridParquetReader
+    lazy val allSupportedTypes = !fsse.requiredSchema.exists { field =>
+      TrampolineUtil.dataTypeExistsRecursively(field.dataType, {
+        // Currently, under some circumstance, the native backend may return incorrect results
+        // over MapType nested by nested types. To guarantee the correctness, disable this pattern
+        // entirely.
+        // TODO: figure out the root cause and support it
+        case ArrayType(_: MapType, _) => true
+        case MapType(_: MapType, _, _) | MapType(_, _: MapType, _) => true
+        case st: StructType if st.exists(_.dataType.isInstanceOf[MapType]) => true
+        // TODO: support DECIMAL with negative scale
+        case dt: DecimalType if dt.scale < 0 => true
+        // TODO: support DECIMAL128
+        case dt: DecimalType if dt.precision > DType.DECIMAL64_MAX_PRECISION => true
+        // TODO: support BinaryType
+        case _: BinaryType => true
+        case _ => false
+      })
+    }
+    // TODO: supports BucketedScan
+    lazy val noBucketedScan = !fsse.bucketedScan
+
+    isEnabled && isParquet && allSupportedTypes && noBucketedScan
+  }
+
+  /**
+   * Check if runtimes are satisfied, including:
+   * - Spark distribution is not CDH or Databricks
+   * - Hybrid jar in the classpath
+   * - Scala version is 2.12
+   * - Parquet V1 data source
+   */
+  def checkRuntimes(v1DataSourceList: String): Unit = {
+    checkNotRunningCDHorDatabricks()
+    HybridExecutionUtils.checkHybridJarInClassPath()
+    checkScalaVersion()
+    checkV1Datasource(v1DataSourceList)
+  }
+
+  /**
+   * Check Spark distribution is not CDH or Databricks,
+   * report error if it is
+   */
+  private def checkNotRunningCDHorDatabricks(): Unit = {
```

Comment:
nit: I would prefer to call these kinds of methods `assert`Something instead of `check`Something. To me it implies more strongly that an exception would be thrown in the wrong case.